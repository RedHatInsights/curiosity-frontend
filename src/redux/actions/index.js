import * as rhelActions from './rhelActions';
import * as userActions from './userActions';

const actions = {
  rhel: rhelActions,
  user: userActions
};

const reduxActions = { ...actions };

export { reduxActions as default, reduxActions, rhelActions, userActions };
