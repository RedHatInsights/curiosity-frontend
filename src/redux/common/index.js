import { reduxHelpers } from './reduxHelpers';

export { reduxHelpers as default, reduxHelpers };
