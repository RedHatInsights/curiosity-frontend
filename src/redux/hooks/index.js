import { reactReduxHooks } from './useReactRedux';

const storeHooks = {
  reactRedux: reactReduxHooks
};

export { storeHooks as default, storeHooks, reactReduxHooks };
