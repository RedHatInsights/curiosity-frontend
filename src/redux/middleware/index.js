import { createLogger } from 'redux-logger';
import promiseMiddleware from 'redux-promise-middleware';
import thunkMiddleware from 'redux-thunk';

const reduxMiddleware = [thunkMiddleware, promiseMiddleware];

if (process.env.NODE_ENV !== 'production' && process.env.REACT_APP_DEBUG_MIDDLEWARE === 'true') {
  reduxMiddleware.push(createLogger());
}

export { reduxMiddleware as default, reduxMiddleware, createLogger, promiseMiddleware, thunkMiddleware };
