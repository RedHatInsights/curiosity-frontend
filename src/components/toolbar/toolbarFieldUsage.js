import React from 'react';
import PropTypes from 'prop-types';
import { reduxTypes, storeHooks } from '../../redux';
import { useProduct, useProductQuery } from '../productView/productViewContext';
import { Select, SelectPosition } from '../form/select';
import { RHSM_API_QUERY_USAGE_TYPES as FIELD_TYPES, RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { translate } from '../i18n/i18n';

/**
 * A standalone Usage select filter.
 *
 * @memberof Toolbar
 * @module ToolbarFieldUsage
 */

/**
 * Select field options.
 *
 * @type {Array<{title: React.ReactNode, value: string, selected: boolean}>}
 */
const toolbarFieldOptions = Object.values(FIELD_TYPES).map(type => ({
  title: translate('curiosity-toolbar.label', { context: ['usage', (type === '' && 'none') || type] }),
  value: type,
  selected: false
}));

/**
 * On select update usage.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnSelect = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { viewId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  return ({ value = null } = {}) =>
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RESET_INVENTORY_LIST,
        viewId
      },
      {
        type: reduxTypes.query.SET_QUERY_RHSM_TYPES[RHSM_API_QUERY_SET_TYPES.USAGE],
        viewId,
        [RHSM_API_QUERY_SET_TYPES.USAGE]: value
      }
    ]);
};

/**
 * Display a usage field with options.
 *
 * @fires onSelect
 * @param {object} props
 * @param {boolean} props.isFilter
 * @param {object} props.options
 * @param {string} props.position
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useProductQuery
 * @returns {React.ReactNode}
 */
const ToolbarFieldUsage = ({
  isFilter,
  options,
  position,
  t,
  useOnSelect: useAliasOnSelect,
  useProductQuery: useAliasProductQuery
}) => {
  const { [RHSM_API_QUERY_SET_TYPES.USAGE]: updatedValue } = useAliasProductQuery();
  const onSelect = useAliasOnSelect();

  const updatedOptions = options.map(option => ({ ...option, selected: option.value === updatedValue }));

  return (
    <Select
      aria-label={t(`curiosity-toolbar.placeholder${(isFilter && '_filter') || ''}`, { context: 'usage' })}
      onSelect={onSelect}
      options={updatedOptions}
      selectedOptions={updatedValue}
      placeholder={t(`curiosity-toolbar.placeholder${(isFilter && '_filter') || ''}`, { context: 'usage' })}
      position={position}
      data-test="toolbarFieldUsage"
    />
  );
};

/**
 * Prop types.
 *
 * @type {{useOnSelect: Function, t: Function, isFilter: boolean, options: Array, useProductQuery: Function,
 *     position: string}}
 */
ToolbarFieldUsage.propTypes = {
  isFilter: PropTypes.bool,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.node,
      value: PropTypes.any,
      selected: PropTypes.bool
    })
  ),
  position: PropTypes.string,
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useProductQuery: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useOnSelect: Function, t: Function, isFilter: boolean, options: Array, useProductQuery: Function,
 *     position: string}}
 */

ToolbarFieldUsage.defaultProps = {
  isFilter: false,
  options: toolbarFieldOptions,
  position: SelectPosition.left,
  t: translate,
  useOnSelect,
  useProductQuery
};

export { ToolbarFieldUsage as default, ToolbarFieldUsage, toolbarFieldOptions, useOnSelect };
