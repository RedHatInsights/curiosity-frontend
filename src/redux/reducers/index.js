import { combineReducers } from 'redux';
import { notifications } from '@redhat-cloud-services/frontend-components-notifications';
import graphReducer from './graphReducer';
import inventoryReducer from './inventoryReducer';
import viewReducer from './viewReducer';
import userReducer from './userReducer';

const reducers = {
  notifications,
  graph: graphReducer,
  inventory: inventoryReducer,
  user: userReducer,
  view: viewReducer
};

const reduxReducers = combineReducers(reducers);

export { reduxReducers as default, reduxReducers, graphReducer, inventoryReducer, userReducer, viewReducer };
