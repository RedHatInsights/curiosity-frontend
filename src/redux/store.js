import { createStore, applyMiddleware } from 'redux';
import { reduxMiddleware } from './middleware';
import { reduxReducers } from './reducers';

const store = createStore(reduxReducers, applyMiddleware(...reduxMiddleware));

export { store as default, store };
