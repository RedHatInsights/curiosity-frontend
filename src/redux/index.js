import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import { store } from './store';
import { reduxActions } from './actions';
import { reduxReducers } from './reducers';
import { reduxSelectors } from './selectors';
import { reduxTypes } from './types';
import { helpers } from '../common/helpers';

const connectTranslate = (mapStateToProps, mapDispatchToProps) => component =>
  connect(
    mapStateToProps,
    mapDispatchToProps
  )((!helpers.TEST_MODE && withTranslation()(component)) || component);

const connectRouterTranslate = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(connectTranslate(mapStateToProps, mapDispatchToProps)(component));

const connectRouter = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(
    connect(
      mapStateToProps,
      mapDispatchToProps
    )(component)
  );

export {
  connect,
  connectRouter,
  connectRouterTranslate,
  connectTranslate,
  reduxActions,
  reduxReducers,
  reduxSelectors,
  reduxTypes,
  store
};
