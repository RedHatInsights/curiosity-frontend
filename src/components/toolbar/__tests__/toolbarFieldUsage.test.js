import React from 'react';
import { shallow } from 'enzyme';
import { ToolbarFieldUsage, toolbarFieldOptions, useOnSelect } from '../toolbarFieldUsage';
import { store } from '../../../redux/store';
import {
  RHSM_API_QUERY_USAGE_TYPES as USAGE_TYPES,
  RHSM_API_QUERY_SET_TYPES
} from '../../../services/rhsm/rhsmConstants';

describe('ToolbarFieldSla Component', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.spyOn(store, 'dispatch').mockImplementation((type, data) => ({ type, data }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a basic component', () => {
    const props = {
      useProductQuery: () => ({ [RHSM_API_QUERY_SET_TYPES.USAGE]: USAGE_TYPES.DEVELOPMENT })
    };
    const component = shallow(<ToolbarFieldUsage {...props} />);

    expect(component).toMatchSnapshot('basic');
  });

  it('should export select options', () => {
    expect(toolbarFieldOptions).toMatchSnapshot('toolbarFieldOptions');
  });

  it('should handle updating usage through redux state with component', async () => {
    const props = {};

    const component = await mountHookComponent(<ToolbarFieldUsage {...props} />);

    component.find('button').simulate('click');
    component.update();
    component.find('button.pf-c-select__menu-item').first().simulate('click');

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch usage, component');
  });

  it('should handle updating usage through redux state with hook', () => {
    const options = {
      useProduct: () => ({ viewId: 'loremIpsum' })
    };

    const onSelect = useOnSelect(options);

    onSelect({
      value: 'dolor sit'
    });
    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch usage, hook');
  });
});
