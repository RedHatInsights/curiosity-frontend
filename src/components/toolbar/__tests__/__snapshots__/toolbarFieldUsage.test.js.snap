// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarFieldSla Component should export select options: toolbarFieldOptions 1`] = `
[
  {
    "selected": false,
    "title": "t(curiosity-toolbar.label_usage, {"context":"Production"})",
    "value": "Production",
  },
  {
    "selected": false,
    "title": "t(curiosity-toolbar.label_usage, {"context":"Development/Test"})",
    "value": "Development/Test",
  },
  {
    "selected": false,
    "title": "t(curiosity-toolbar.label_usage, {"context":"Disaster Recovery"})",
    "value": "Disaster Recovery",
  },
  {
    "selected": false,
    "title": "t(curiosity-toolbar.label_usage, {"context":"none"})",
    "value": "",
  },
]
`;

exports[`ToolbarFieldSla Component should handle updating usage through redux state with component: dispatch, component 1`] = `
[
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "type": "SET_QUERY_RHSM_usage",
        "usage": "Production",
        "viewId": undefined,
      },
    ],
  ],
]
`;

exports[`ToolbarFieldSla Component should handle updating usage through redux state with hook: dispatch, hook 1`] = `
[
  [
    [
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": "loremIpsum",
      },
      {
        "type": "SET_QUERY_RHSM_usage",
        "usage": "dolor sit",
        "viewId": "loremIpsum",
      },
    ],
  ],
]
`;

exports[`ToolbarFieldSla Component should render a basic component: basic 1`] = `
<Select
  aria-label="t(curiosity-toolbar.placeholder, {"context":"usage"})"
  ariaLabel="Select option"
  className=""
  data-test="toolbarFieldUsage"
  direction="down"
  id="generatedid-"
  isDisabled={false}
  isDropdownButton={false}
  isFlipEnabled={false}
  isInline={true}
  isToggleText={true}
  maxHeight={null}
  name={null}
  onSelect={[Function]}
  onSplitButton={[Function]}
  options={
    [
      {
        "selected": false,
        "title": "t(curiosity-toolbar.label_usage, {"context":"Production"})",
        "value": "Production",
      },
      {
        "selected": true,
        "title": "t(curiosity-toolbar.label_usage, {"context":"Development/Test"})",
        "value": "Development/Test",
      },
      {
        "selected": false,
        "title": "t(curiosity-toolbar.label_usage, {"context":"Disaster Recovery"})",
        "value": "Disaster Recovery",
      },
      {
        "selected": false,
        "title": "t(curiosity-toolbar.label_usage, {"context":"none"})",
        "value": "",
      },
    ]
  }
  placeholder="t(curiosity-toolbar.placeholder, {"context":"usage"})"
  position="left"
  selectedOptions="Development/Test"
  splitButtonVariant={null}
  toggleIcon={null}
  variant="single"
/>
`;
